File: src/csharp/ejercicios/Ejercicio4.cs
using System;
using System.Threading;

class Ejercicio4
{
    public static void Run()
    {
        int contadorInseguro = 0;
        int contadorSincronizado = 0;
        var lockObj = new object();

        // SIN sincronizacion
        var hilos = new Thread[5];
        for (int i = 0; i < 5; i++)
        {
            hilos[i] = new Thread(() => {
                for (int j = 0; j < 10000; j++)
                    contadorInseguro++;
            });
            hilos[i].Start();
        }
        foreach (var h in hilos) h.Join();
        Console.WriteLine($"Esperado: 50000 | Sin control: {contadorInseguro}");

        // CON lock
        for (int i = 0; i < 5; i++)
        {
            hilos[i] = new Thread(() => {
                for (int j = 0; j < 10000; j++)
                    lock (lockObj) { contadorSincronizado++; }
            });
            hilos[i].Start();
        }
        foreach (var h in hilos) h.Join();
        Console.WriteLine($"Esperado: 50000 | Con lock: {contadorSincronizado}");

        // CON Interlocked
        int contadorInterlocked = 0;
        for (int i = 0; i < 5; i++)
        {
            hilos[i] = new Thread(() => {
                for (int j = 0; j < 10000; j++)
                    Interlocked.Increment(ref contadorInterlocked);
            });
            hilos[i].Start();
        }
        foreach (var h in hilos) h.Join();
        Console.WriteLine($"Esperado: 50000 | Con Interlocked: {contadorInterlocked}");
    }
}

File: src/python/ejercicios/Ejercicio4.py
import threading

contador_inseguro = 0
contador_sincronizado = 0
lock = threading.Lock()

def incrementar_inseguro():
    global contador_inseguro
    for _ in range(10000):
        contador_inseguro += 1

def incrementar_sincronizado():
    global contador_sincronizado
    for _ in range(10000):
        with lock:
            contador_sincronizado += 1

# SIN sincronizacion
hilos = [threading.Thread(target=incrementar_inseguro) for _ in range(5)]
for h in hilos: h.start()
for h in hilos: h.join()
print(f"Esperado: 50000 | Sin control: {contador_inseguro}")

# CON Lock
hilos = [threading.Thread(target=incrementar_sincronizado) for _ in range(5)]
for h in hilos: h.start()
for h in hilos: h.join()
print(f"Esperado: 50000 | Con Lock: {contador_sincronizado}")

